공지사항 등록 API 개발

zlzzlzz2l opened the pull request:
# 공지사항 등록 API 개발

---

공지사항 등록 API를 추가하였습니다.

---
올바른 형식으로 공지사항을 등록했을 때
![image](https://user-images.githubusercontent.com/48669011/160212061-801f4e56-ce37-4fc3-ba45-1fe93b859313.png)

제목 혹은 본문이 빈칸일 때
![image](https://user-images.githubusercontent.com/48669011/160212088-eacdf6a1-ddd0-4f9f-a21b-94ba39a892c7.png)

![image](https://user-images.githubusercontent.com/48669011/160212109-115d8ba1-f501-4979-86e8-e5aed73189bf.png)

zlzzlzz2l committed:
Update: #32 - 공지사항 등록 메시지 추가

zlzzlzz2l committed:
Update: #32 - 공지사항 등록 메시지 static 변수 추가

zlzzlzz2l committed:
Style: #32 - 에러 메시지 오타 수정

zlzzlzz2l committed:
Update: #32 - 공지사항 등록 에러 메시지 추가

zlzzlzz2l committed:
Update: #32 - 공지사항 등록 에러 메시지 static 변수 추가

zlzzlzz2l committed:
Update: #32 - 공지사항 등록 요청 DTO 추가

- String 타입 title(제목)과 content(본문)을 선언하였습니다.
- 제목과 본문은 null, "", " "을 허용하지 않기 때문에 NotBlank 어노테이션 추가하였습니다.

zlzzlzz2l committed:
Update: #32 - 공지사항 컨트롤러 생성 및 공지사항 등록 메서드 추가

- 관리자만 공지사항을 등록할 수 있기 때문에 관리자만 접근할 수 있도록 설정하였습니다.
- HTTP 응답 코드가 1개이기 때문에 ResponseStatus 어노테이션을 적용하였습니다.
- 인증된 사용자 정보에서 seq를 가져오기 위해 Long 타입의 memberSeq을 선언하였습니다.
- 공지사항 등록 요청 DTO와 사용자 seq를 인자로 가진 공지사항 등록 메서드를 호출합니다.

zlzzlzz2l committed:
Update: #32 - 공지사항 서비스 인터페이스 생성 및 등록, 삭제 메서드 선언부 추가

zlzzlzz2l committed:
Update: #32 - 회원 레포지토리 메서드 선언부 추가

- 식별자로 회원을 찾기 위해 해당 메서드를 추가하였습니다.

zlzzlzz2l committed:
Update: #32 - 공지사항 서비스 구현 클래스 생성 및 등록, 삭제 메서드 구현

- memberSeq 매개변수를 이용하여 회원을 조회합니다.
- 객체 생성 후 save 메서드를 통해 저장합니다.

zlzzlzz2l committed:
Update: #32 - 공지사항 레포지토리 메서드 선언부 추가

- 식별자로 공지사항을 찾기 위해 추가하였습니다.

zlzzlzz2l committed:
Test: #32 - 공지사항 서비스 테스트 공지사항 등록 메서드 추가

- 공지사항을 등록한 후 식별자를 이용하여 공지사항을 조회합니다.
- 공지사항 요청 DTO와 저장된 공지사항을 비교합니다.

zlzzlzz2l committed:
Test: #32 - 공지사항 레포지토리 테스트 엔티티 등록 메서드 추가

- 공지사항을 등록한 후 식별자를 이용하여 공지사항을 조회합니다.
- 공지사항 엔티티와 저장된 공지사항을 비교합니다.

0xe82de commented on src/main/java/com/fourfourfour/damoa/api/notice/repository/NoticeRepository.java:
```diff
@@ -0,0 +1,12 @@
+
+public interface NoticeRepository extends JpaRepository<Notice, Long> {
+
+    @Query("SELECT n FROM Notice n WHERE n.seq = :seq")
+    Notice findBySeq(@Param("seq") Long seq);
```
```java
@Query("SELECT n FROM Notice n WHERE n.seq = :seq AND n.isDeleted = FALSE")
Notice findBySeq(@Param("seq") Long seq);
```

공지사항 데이터는 `isDeleted` 컬럼으로 삭제 여부를 관리하기 때문에 위처럼 바꾸는 것이 좋은 것 같아요 :)

zlzzlzz2l replied:
아하 넵넵 반영했습니다!

zlzzlzz2l replied:
임시로 DB에 값을 추가하고 테스트했습니다! 
아래 그림과 같이 DELETE 쿼리가 정상적으로 나오는 것을 확인할 수 있습니다.

![image](https://user-images.githubusercontent.com/48669011/160535448-a6a41882-37d2-4557-8bf3-6111fdc8638e.png)

![image](https://user-images.githubusercontent.com/48669011/160535255-9d34dd7b-c3e7-4303-9188-25ef740a5ec1.png)
![image](https://user-images.githubusercontent.com/48669011/160535278-41d81b14-0a58-4365-9a54-b46494c3a637.png)
![image](https://user-images.githubusercontent.com/48669011/160535316-0f533581-9b33-45ce-afa4-6fa33b24a0dd.png)

0xe82de commented on src/test/java/com/fourfourfour/damoa/api/notice/service/NoticeServiceImplTest.java:
```diff
@@ -0,0 +1,83 @@
+    @BeforeEach
+    public void setUp() {
+
+        memberService.deleteAll();
+        noticeService.deleteAll();
+
```
이렇게 했을 때 `delete` 쿼리가 정상적으로 나가나요?? db에 값이 있을 때요.

zlzzlzz2l replied:
임시로 DB에 값을 추가하고 테스트했습니다!
아래 그림과 같이 `DELETE` 쿼리가 정상적으로 나오는 것을 확인할 수 있습니다.

![image](https://user-images.githubusercontent.com/48669011/160408625-9e89a518-591a-4bd2-9887-22bc11122e9c.png)

![image](https://user-images.githubusercontent.com/48669011/160408840-240b6344-5f78-47b7-a9d3-45e93822b16b.png)
![image](https://user-images.githubusercontent.com/48669011/160408916-5c1b2c8d-78c8-43f6-ade3-0b8518366ab6.png)
![image](https://user-images.githubusercontent.com/48669011/160408950-291ba279-804b-43fd-a857-efe36f8453ce.png)

zlzzlzz2l replied:
임시로 DB에 값을 추가하고 테스트했습니다! 
아래 그림과 같이 DELETE 쿼리가 정상적으로 나오는 것을 확인할 수 있습니다.

![image](https://user-images.githubusercontent.com/48669011/160535448-a6a41882-37d2-4557-8bf3-6111fdc8638e.png)

![image](https://user-images.githubusercontent.com/48669011/160535255-9d34dd7b-c3e7-4303-9188-25ef740a5ec1.png)
![image](https://user-images.githubusercontent.com/48669011/160535278-41d81b14-0a58-4365-9a54-b46494c3a637.png)
![image](https://user-images.githubusercontent.com/48669011/160535316-0f533581-9b33-45ce-afa4-6fa33b24a0dd.png)

0xe82de commented on src/test/java/com/fourfourfour/damoa/api/notice/repository/NoticeRepositoryTest.java:
```diff
@@ -0,0 +1,82 @@
+                .writer(member)
+                .build();
+
+        noticeRepository.save(notice);
+
```
`em.flush()`, `em.clear()` 추가해야되지 않을까 싶습니다.

zlzzlzz2l replied:
넵 반영했습니다.

0xe82de reviewed (changes requested):
고생하셨습니다. 코멘트 확인해주시고 피드백 주세요 :)

zlzzlzz2l committed:
Refactor: #32 - 공지사항 레포지토리 메서드 선언부 수정

- 식별자와 삭제 여부로 공지사항을 찾기 위해 수정하였습니다.

zlzzlzz2l committed:
Refactor: #32 - 공지사항 레포지토리 테스트 엔티티 등록 메서드 수정

- 영속성 컨텍스트 안에 있는 데이터를 DB로 보내고나서 비우기 위해 flush와 clear를 해주었습니다.
- member 객체와 프록시 객체로 비교하기 때문에 식별자를 이용하여 검증하였습니다.

0xe82de reviewed (approved):
고생하셨습니다~

yjs1213 reviewed (approved):
고생하셨어요
